### PR TITLE
[WIP] Codechain

### PR DESCRIPTION
**Pull request by MaxHillebrand**
This is a work in progress branch of documentation about code chain, a tool to prevent targeted backdoors, with a way to sign and distribute source code and compiled packages with m-of-n review signatures in key rotation and single source of truth.

TODO:
- [x] rational with the perceived problem and proposed solution
- [ ] summary of the protocol design
- [ ] step by step tutorial for code signers
- [ ] step by step guide for users
- [ ] FAQ

This should only be merged if we end up using codechain in Wasabi. If not, then it might be forked to another repo to keep the information for other projects who might use it.

### CODE REVIEW

File: docs/building-wasabi/CodeChain.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+---
+{
+  "title": "CodeChain",
+  "description": "A guide on how to use Code Chain to sign and distribute the Wasabi source and package. This is the Wasabi documentation, an archive of knowledge about the open-source, non-custodial and privacy-focused Bitcoin wallet for desktop."
```
Comment (yahiheb) [inline docs/building-wasabi/CodeChain.md]:
```suggestion
  "description": "A guide on how to use CodeChain to sign and distribute the Wasabi source and package. This is the Wasabi documentation, an archive of knowledge about the open-source, non-custodial and privacy-focused Bitcoin wallet for desktop."
```

File: docs/building-wasabi/CodeChain.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+### The Problem
+
+For high security software like Wasabi, we need to take extra care that the code is well reviewed by different peers, and that users run exactly that software, and are not victim of a targeted backdoor attack.
+Current best practice is to have PGP signatures of the developers on the individual code commits, as well as by the maintainers on every released compiled package.
+However, users do not verify the git commit signatures, and only a few verify the build signatures.
+Packages are distributed on the Wasabi website, with no verification of what the most recent version is.
+
+There are several problems with such an approach:
+- The PGP keys can be very and this long validity time makes stolen keys disastrous.
```
[the remark this answers is not in the document]
Comment (MaxHillebrand) [reply docs/building-wasabi/CodeChain.md]: old...

**Comment by nixbitcoin** [comment]
I'm researching CodeChain for [nix-bitcoin](https://github.com/fort-nix/nix-bitcoin). I saw that you have some previous work here, so I decided to put my work on top of yours. I've added a summary of the protocol design on https://github.com/nixbitcoin/WasabiDoc/commits/codechain. 

If this gets engagement, I'm open to completing a "step by step tutorial for code signers" and a "step by step guide for users" based on a sample project.

**Comment by MaxHillebrand** [comment]
this is a great summary @nixbitcoin, thanks for writing it!

FYI, this initial PR was a very very early PoC, with no intent to get it merged, at least until Wasabi implements CodeChain, which there is no ongoing effort to do.

So, feel free to use this content in any shape for wherever you want to publish it!

**Comment by nopara73** [comment]
It has nothing to do with Wasabi.

**Pull request closed without merging**